File: apps/control-plane/internal/logsafe/logsafe.go
// Package logsafe sanitises untrusted values before they are written to logs.
//
// Log injection (CodeQL go/log-injection): any attacker-controlled string that
// reaches a log line can contain CR/LF and forge additional log entries. An
// attacker who can pick their own email address, OAuth provider name, or
// webhook id can therefore write arbitrary lines into our logs — enough to
// fake an audit trail, hide a real entry in noise, or break a downstream log
// parser that splits on newlines.
//
// Sanitising at the log call site (rather than at input validation) is
// deliberate: these values are legitimately allowed to contain odd characters
// elsewhere in the system, and we do not want to reject a signup because an
// address is unusual. The log line is the only place the bytes are dangerous.
package logsafe

import "strings"

// Max length for a sanitised value. Long enough for an email, a UUID, or a
// provider id; short enough that a megabyte of junk cannot flood a log file.
const maxLen = 256

// String returns v with anything that could forge a log entry removed:
// newlines, carriage returns, tabs, and other C0/C1 control characters. The
// result is truncated to a sane length and marked when truncated.
//
// Replacement (rather than deletion) keeps the value's shape visible, so a
// forged-looking entry is obvious in the log rather than silently reassembled.
func String(v string) string {
	truncated := false
	if len(v) > maxLen {
		v = v[:maxLen]
		truncated = true
	}

	var b strings.Builder
	b.Grow(len(v))
	for _, r := range v {
		switch {
		case r == '\n' || r == '\r':
			// The characters that actually forge a new log entry.
			b.WriteString("\\n")
		case r == '\t':
			b.WriteString("\\t")
		case r < 0x20 || r == 0x7f:
			// Remaining C0 controls + DEL: terminal escapes, NUL, etc.
			b.WriteByte('?')
		default:
			b.WriteRune(r)
		}
	}

	out := b.String()
	if truncated {
		out += "…(truncated)"
	}
	return out
}

File: apps/web/src/lib/log-safe.ts
/**
 * Sanitises untrusted values before they are written to server logs.
 *
 * Log injection (CodeQL js/log-injection): any user-controlled string that
 * reaches a log line can contain CR/LF and forge additional entries. On a
 * feedback endpoint the whole payload is attacker-chosen, so an unescaped title
 * or body can write arbitrary lines into the server log — enough to fake an
 * entry, bury a real one, or break a parser that splits on newlines.
 *
 * Sanitising at the log call site is deliberate: the same text is legitimately
 * allowed to contain newlines when it becomes a GitHub issue body. Only the log
 * line is dangerous, so only the log line is escaped.
 */

/** Long enough for a feedback title or a stack trace line; short enough that a
 * megabyte of junk cannot flood the log. */
const MAX_LEN = 2000;

export function logSafe(value: unknown): string {
  let text = typeof value === "string" ? value : String(value);

  let truncated = false;
  if (text.length > MAX_LEN) {
    text = text.slice(0, MAX_LEN);
    truncated = true;
  }

  // Replace rather than delete: the value's shape stays visible, so a
  // forged-looking entry reads as escaped text instead of silently
  // reassembling into something that looks like a real log line.
  const escaped = text
    .replace(/\r\n|\r|\n/g, "\\n")
    .replace(/\t/g, "\\t")
    // Remaining C0 controls + DEL: terminal escapes, NUL, etc.
    // biome-ignore lint/suspicious/noControlCharactersInRegex: matching control characters is the entire point
    .replace(/[\u0000-\u001F\u007F]/g, "?");

  return truncated ? `${escaped}…(truncated)` : escaped;
}
